File: app/src/main/java/com/example/uos_lms/core/domain/model/LeaveApplication.java
package com.example.uos_lms.core.domain.model;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@Builder(toBuilder = true)
@EqualsAndHashCode
@ToString
public class LeaveApplication {
    @Builder.Default
    private final String id = "";
    @Builder.Default
    private final String studentUid = "";
    @Builder.Default
    private final String studentName = "";
    @Builder.Default
    private final String departmentId = "";
    @Builder.Default
    private final String semesterId = "";
    private final long fromDateMillis;
    private final long toDateMillis;
    @Builder.Default
    private final String reason = "";
    private final String attachmentUrl;
    private final String attachmentName;
    private final String attachmentPublicId;
    private final String attachmentResourceType;
    private final Long attachmentSize;
    @Builder.Default
    private final LeaveStatus status = LeaveStatus.PENDING;
    private final String reviewerUid;
    private final String reviewerName;
    private final String reviewerRole;
    private final Long decidedAt;
    @Builder.Default
    private final long createdAt = 0L;
}

File: src/models/LeaveApplication.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const leaveApplicationSchema = new Schema(
  {
    studentId: { type: String, ref: 'User', required: true },
    departmentId: { type: Schema.Types.ObjectId, ref: 'Department', required: true },
    semesterId: { type: Schema.Types.ObjectId, ref: 'Semester', default: null },

    fromDate: { type: Date, required: true },
    toDate: { type: Date, required: true },
    reason: { type: String, required: true },

    mediaId: { type: Schema.Types.ObjectId, ref: 'Media', default: null },
    attachmentUrl: { type: String, default: null },
    attachmentName: { type: String, default: null },
    attachmentPublicId: { type: String, default: null },
    attachmentResourceType: { type: String, default: null },
    attachmentSize: { type: Number, default: null },

    status: { type: String, enum: ['PENDING', 'APPROVED', 'REJECTED'], default: 'PENDING' },
    reviewerId: { type: String, ref: 'User', default: null },
    reviewerRole: { type: String, default: null },
    decidedAt: { type: Date, default: null },
  },
  { timestamps: true }
);

module.exports = mongoose.model('LeaveApplication', leaveApplicationSchema);
